File: data_structures/03-stack/stack.cpp
#include "stack.hpp"
#include <stdexcept>

template <typename T>
Stack<T>::Stack() : Container<T>(), top_() {}


template <typename T>
int Stack<T>::size() const {
    return this->size_;
}

template <typename T>
bool Stack<T>::empty() const {
    return this->size_ == 0;
}


template <typename T>
T Stack<T>::top() const {
    if (!top_) {
        throw std::out_of_range("Stack<T>::top() : index out of range");
    }
    return top_->data;
}

template <typename T>
void Stack<T>::push(T value) {
    Node* new_node = new Node(value);
    new_node->next = top_;
    top_ = new_node;
}

template <typename T>
T Stack<T>::pop() {
    if (!top_) {
        throw std::out_of_range("Stack<T>::top() : index out of range");
    }

    Node* old_top = top_;
    T top_value = old_top->data;

    top_ = top_->next;
    delete old_top;
}


template <typename T>
std::string Stack<T>::ToString() const {
    std::string s;
    for (Node* ptr = top_; ptr; ptr = ptr->next) {
        s += std::to_string(ptr->data) + " ";
    }
    return s;
}

File: Snippets/java/0-Syntax/Stack.java
class Node<T> {

    private T _data;
    private Node<T> _next;

    public Node(T data) {
        _data = data;
        _next = null;
    }

    public T getData() {
        return _data;
    }

    public Node<T> getNext() {
        return _next;
    }

    public void setNext(Node<T> next) {
        _next = next;
    }

}


public class Stack<T> {
    
    private Node<T> _top;

    public Stack() {
        _top = null;
    }
    
    public void push(T data) {
        if (isEmpty()) {
            _top = new Node<T>(data);
        } else {
            Node<T> ptr = new Node<T>(data);
            ptr.setNext(_top);
            _top = ptr;
        }
    }

    public T pop() {
        if (isEmpty()) {
            return null;
        } else {
            Node<T> ptr = _top;
            _top = _top.getNext();
            return ptr.getData();
        }
    }

    public boolean isEmpty() {
        return (_top == null);
    }

    @Override
    public String toString() {
        String ret = "[";
        Node<T> ptr = _top;
        while (ptr != null) {
            ret = ptr.getData().toString();
        }
        return ret + "]";
    }


    public static void main(String[] args) {
        Stack<String> s = new Stack<>();
        s.push("world");
        s.push("hello");
        System.out.println(s.pop());
        System.out.println(s.pop());
    }
}
